x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -156,18 +160,19 @@ public Void run() {
     // private final Environment env;
     protected final Licensing licensing;
     // These should not be directly accessed as they cannot be overridden in tests. Please use the getters so they can be overridden.
-    private static final SetOnce<XPackLicenseState> licenseState = new SetOnce<>();
     private static final SetOnce<SSLService> sslService = new SetOnce<>();
-    private static final SetOnce<LicenseService> licenseService = new SetOnce<>();
     private static final SetOnce<LongSupplier> epochMillisSupplier = new SetOnce<>();
+    // non-final to allow for testing
+    private static SetOnce<XPackLicenseState> licenseState = new SetOnce<>();
+    private static SetOnce<LicenseService> licenseService = new SetOnce<>();
 
     public XPackPlugin(final Settings settings) {
         super();
         // FIXME: The settings might be changed after this (e.g. from "additionalSettings" method in other plugins)
         // We should only depend on the settings from the Environment object passed to createComponents
         this.settings = settings;
-
-        setLicenseState(new XPackLicenseState(() -> getEpochMillisSupplier().getAsLong()));
+        licenseState = new SetOnce<>();
+        licenseService = new SetOnce<>();
 
```

Comment:
Alternatively, we could have the fields to be package private so that they can be reset by the tests. This helps confining the trade-off in a single place (the fields) and avoid having to re-initialize them in the constructor. I don't think it's perfect either. But would like to offer it for your consideration.

Reply:
If i moved the package private I would still to make them non-final, resulting in non-final and increased visibility. By initializing them to the constructor I still need to make non-final, but can keep the as private.